File: java/src/main/java/io/cucumber/prettyformatter/StepTextFormatter.java
package io.cucumber.prettyformatter;

import io.cucumber.messages.types.Group;
import io.cucumber.messages.types.PickleStep;
import io.cucumber.messages.types.StepMatchArgument;
import io.cucumber.messages.types.TestStep;

import java.util.ArrayList;
import java.util.List;

import static io.cucumber.prettyformatter.Theme.Element.STEP_ARGUMENT;
import static io.cucumber.prettyformatter.Theme.Element.STEP_TEXT;
import static java.util.Collections.emptyList;

final class StepTextFormatter {

    StepTextFormatter() {
    }

    void formatTo(TestStep testStep, PickleStep pickleStep, LineBuilder lineBuilder) {
        formatStepText(lineBuilder, pickleStep.getText(), getStepMatchArguments(testStep));
    }

    private List<StepMatchArgument> getStepMatchArguments(TestStep testStep) {
        List<StepMatchArgument> stepMatchArguments = new ArrayList<>();
        testStep.getStepMatchArgumentsLists()
                .filter(stepMatchArgumentsList -> stepMatchArgumentsList.size() == 1)
                .orElse(emptyList())
                .forEach(list -> stepMatchArguments.addAll(list.getStepMatchArguments()));
        return stepMatchArguments;
    }

    private void formatStepText(LineBuilder lineBuilder, String stepText, List<StepMatchArgument> arguments) {
        int currentIndex = 0;
        for (StepMatchArgument argument : arguments) {
            Group group = argument.getGroup();
            // Ignore absent values, or groups without a start
            if (group.getValue().isPresent() && group.getStart().isPresent()) {
                String groupValue = group.getValue().get();
                // TODO: Messages are silly
                int groupStart = (int) (long) group.getStart().get();
                String text = stepText.substring(currentIndex, groupStart);
                currentIndex = groupStart + groupValue.length();
                lineBuilder.append(STEP_TEXT, text)
                        .append(STEP_ARGUMENT, groupValue);
            }
        }
        if (currentIndex != stepText.length()) {
            String remainder = stepText.substring(currentIndex);
            lineBuilder.append(STEP_TEXT, remainder);
        }
    }
}

File: cpp/src/cucumber/pretty-formatter/StepTextFormatter.cpp
#include "cucumber/pretty-formatter/StepTextFormatter.hpp"
#include "cucumber/messages/PickleStep.hpp"
#include "cucumber/messages/StepMatchArgument.hpp"
#include "cucumber/messages/TestStep.hpp"
#include "cucumber/pretty-formatter/LineBuilder.hpp"
#include "cucumber/pretty-formatter/Theme.hpp"
#include <cstddef>
#include <string>
#include <vector>

namespace cucumber::pretty_formatter
{
    void StepTextFormatter::Format(LineBuilder& lineBuilder, const messages::TestStep& testStep,
        const messages::PickleStep& pickleStep) const
    {
        Format(lineBuilder, pickleStep.text, GetStepMatchArguments(testStep));
    }

    std::vector<const messages::StepMatchArgument*> StepTextFormatter::GetStepMatchArguments(const messages::TestStep& testStep) const
    {
        std::vector<const messages::StepMatchArgument*> stepMatchArguments;

        if (testStep.stepMatchArgumentsLists && testStep.stepMatchArgumentsLists->size() == 1)
        {
            for (const auto& argumentsList : *testStep.stepMatchArgumentsLists)
            {
                for (const auto& stepMatchArgument : argumentsList.stepMatchArguments)
                {
                    stepMatchArguments.push_back(std::addressof(stepMatchArgument));
                }
            }
        }

        return stepMatchArguments;
    }

    void StepTextFormatter::Format(LineBuilder& lineBuilder, std::string_view stepText,
        const std::vector<const messages::StepMatchArgument*>& stepMatchArguments) const
    {
        std::size_t pos = 0;

        for (const auto& stepArgument : stepMatchArguments)
        {
            const auto& group = stepArgument->group;
            if (group.value && group.start)
            {
                const auto groupValue = *group.value;
                const auto groupStart = *group.start;
                const auto text = stepText.substr(pos, groupStart - pos);

                pos = groupStart + groupValue.size();

                lineBuilder.Append(Theme::Element::stepText, text).Append(Theme::Element::stepArgument, groupValue);
            }
        }

        if (const auto size = stepText.size(); pos != size)
        {
            lineBuilder.Append(Theme::Element::stepText, stepText.substr(pos));
        }
    }
}
